File: android_app/lib/services/user_level_service.dart
import 'dart:convert';
import 'package:android_app/providers/uri_provider.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:http/http.dart' as http;
import 'package:android_app/models/user_level_info.dart';
import 'package:android_app/services/auth_service.dart';

class UserLevelService {
  final String baseUrl;
  final AuthService authService;

  UserLevelService(this.baseUrl, this.authService);

  Future<UserExperienceInfo> getUserExpInfo() async {
    final response = await http.get(
      Uri.parse('$baseUrl/accounts/me/exp-info'),
      headers: {
        'Content-Type': 'application/json',
        'x-session-id': authService.getSessionId(),
      },
    );

    if (response.statusCode == 200) {
      return UserExperienceInfo.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load user experience info');
    }
  }

  Future<UserExperienceInfo?> getUserExpInfoById(String id) async {
    final response = await http.get(
      Uri.parse('$baseUrl/accounts/id/$id/exp-info'),
      headers: {
        'Content-Type': 'application/json',
        'x-session-id': authService.getSessionId(),
      },
    );

    if (response.statusCode == 200) {
      return UserExperienceInfo.fromJson(jsonDecode(response.body));
    } else {
      return null;
    }
  }
}

final userLevelServiceProvider = Provider<UserLevelService>((ref) {
  final baseUrl = ref.read(apiUrlProvider);
  final authService = ref.read(authServiceProvider);
  return UserLevelService(baseUrl, authService);
});

final userLevelProvider = FutureProvider<UserExperienceInfo?>((ref) async {
  final service = ref.read(userLevelServiceProvider);
  return await service.getUserExpInfo();
});

final userLevelByIdProvider =
    FutureProvider.family<UserExperienceInfo?, String>((ref, id) async {
  final service = ref.read(userLevelServiceProvider);
  return await service.getUserExpInfoById(id);
});

File: client/src/app/services/user-level.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from 'src/environments/environment';
import { UserAuthenticationService } from './user-authentication.service';
import { UserExperienceInfo } from '@common/interfaces/user-experience';

@Injectable({
    providedIn: 'root',
})
export class UserLevelService {
    constructor(
        private readonly httpClient: HttpClient,
        private readonly authService: UserAuthenticationService,
    ) {}

    getUserExpInfo() {
        return this.httpClient.get<UserExperienceInfo>(`${environment.serverUrl}/accounts/me/exp-info`, {
          headers: this.authService.getSessionIdHeaders(),
          observe: 'body',
          responseType: 'json',
        });
      }
      
      getUserExpInfoById(id: string) {
        return this.httpClient.get<UserExperienceInfo>(`${environment.serverUrl}/accounts/id/${id}/exp-info`, {
            headers: this.authService.getSessionIdHeaders(),
            observe: 'body',
            responseType: 'json',
        });
    }
    
      
}
